host/sgx/create.c

```diff
@@ -676,37 +738,52 @@ static oe_result_t _eeid_resign(
     return result;
 }
 
-static oe_result_t _add_eeid_pages(
+static oe_result_t _add_eeid_data_pages(
     oe_sgx_load_context_t* context,
-    uint64_t enclave_addr,
+    oe_enclave_t* enclave,
+    oe_sgx_enclave_properties_t* props,
+    uint64_t entry,
     uint64_t* vaddr)
 {
     oe_result_t result = OE_UNEXPECTED;
-    oe_eeid_t* eeid = context->eeid;
 
-    if (eeid)
+    /* Note: context->eeid_setting is also needed at signing time, for static
```

Comment:
Why does the "static size" EEID enclave's base image needs context->eeid_setting?

Reply:
To distinguish it from non-EEID images. If EEID is enabled at signing time, it will leave a gap/guard page, while the "normal" non-EEID behavior is unchanged.

Reply:
The design goal is to add the guard page between the Code and Initialized Data pages and the heap pages in non-EEID enclave too. Once the design proposal is adopted, we can simplify the implementaiton with regard to some of the "OE_WITH_EXPERIMENTAL_EEID" compiling switches.

Reply:
Right, I thought you wanted to keep the existing behavior unchanged. But, even then, the enclave has no way to tell whether it should expect a populated EEID page. In the past I experimented with a magic value like 0xEE1D on the unmeasured marker page, but that isn't great either. Currently, the `_eeid_rva` symbol is patched with the offset of the EEID page and a non-EEID enclave gets a `0` in there. For this to work, I have to know whether EEID is enabled at signing time though, otherwise the base image hash (and hash state) would change.

Reply:
Based on the updated design, _eeid_rva (or the addr of the guard page before the heap, in the regualr enclaves) is fixed as (_heap_rva - 4KB). _heap_rva is already part of the intialized data of the enclave and can be accessed by the enclave code. For the "static size" EEID enclaves, the size of the enclave does not need to be patched ("dyanmic size" EEID enclave needs to patch the size as a fixed, large number), neither does the entry_point in enclave property (the updated deisgn calls for the regaulr enclave to include entry_point in enclave ptoperty too). The build tool does not need to distinuish "static size" EEID enclave vs. regualr enclave. "Dynamic size" EEID enlcave does have to be "patched" during build time by the build tool.

Reply:
We can't patch any base image at creation time, regardless of whether its statically or dynamically patched, because the verifier would have to re-patch it to verify the evidence, which in turn also means you have to have a copy of the image so you can patch it. Any patching must be done during signing and in exactly the same with (with the same numbers) during enclave creation. (Just in case: "patching" means modifying the base image in any way.)

You listed only the easy cases: what about non-EEID enclaves, how do they know not to expect an EEID page? Should they try to read from the page and then somehow catch the access violation?

I don't know what your "build tool" is, but `oesign` definitely needs to distinguish between EEID and non-EEID images, except if we find a way to distinguish them later from inside the enclave. This was easy in the past, because all images had their memory sizes set to 0/0/1, but that's not the case with statically sized images anymore.

Reply:
I realized that `_eeid_rva` is measured as part of the base image and used to indicate whether the base image is signed as an EEID enclave. I'm OK with this design.

Reply:
Outdated.